tasks.1
.TH TASKS 1 "March 13 2021" "Amchik" "User Commands"
.SH NAME
tasks \- Simple ToDo Manager
.SH SYNOPSIS
.B tasks
.br
.B tasks
[\fI\,token [\fI\,argument\/\fR]...\/\fR]...
.SH DESCRIPTION

Tasks (Amchik/tasks) is a simple ToDo manager written
in C99. Have own config (taskrc or Tasksfile) format,
that simple and can be read without formatting.

.TP
.B "help"
.br
Show a very base documentaion.

.TP
.B "use" <\fI\,file\/\fR>
.br
Select a file to use. For select \fB\,~/.taskrc\/\fR
instead \fB\,Tasksfile\/\fR (by default) use \fB\,%\/\fR.

.TP
.B "table"
.br
Prints a pretty table by the file.

.TP
.B "diagnostic"
.br
Show a errors in file.

.TP
.B "edit" <\fI\,type\/\fR> <\fI\,primary\/\fR> <\fI\,key\/\fR> <\fI\,value\/\fR> 
.br
Edit a record of type \fI\,type\/\fR with primary
key \fI\,primary\/\fR. In type \fItask\fR you can
pass \fB@\fINUM\fR, where NUM is number of task in
table. See section \fBEDITING TASKRC\fR for more info.

.TP
.B "delete" <\fI\,type\/\fR> <\fI\,primary\/\fR>
.br
Delete a record of type \fI\,type\/\fR with primary
key \fI\,primary\/\fR. Like \fBedit\fR in type
\fItask\fR you can pass \fB@\fINUM\fR, where NUM
is number of task in table.

.TP
.B "new" <\fIquery\fR>
.br
Appends a \fIquery\fR to taskrc. For more info see
section \fBTASKRC\fR.

.PP

Running tasks without arguments equals running with
only \fBtable\fR argument.

.SH TASKRC

TaskRC are config type, using by tasks. By default tasks
use \fBTasksfile\fR file in current directory, if it not
exists using \fB~/.taskrc\fR

.SS Syntax
.EX
<\fItype\fR> <\fIprimary\fR> [\fBwith\fR <<\fIkey\fR> <\fIvalue\fR>>[, <\fIkey\fR> <\fIvalue\fR>]... [\fBand\fR <\fIkey\fR> <\fIvalue\fR>]] [\fIpost\fR]
.EE

Type \fBtask\fR contains 2 keys: \fBlabel\fR and \fBpriority\fR.
Set post to \fBcompleted\fR for complete task.

Type \fBlabel\fR have only \fBcolor\fR key, formatted as
\fB#HHHHHH\fR. Note: color #000000 will be not displayed.

.SS Examples

.EX
# Comment!
task "Fix this \\"todo manager\\"" with label github and priority 0  
task "Make a soup"               with label food   ,   priority 10 completed
task "Go to a shop"              completed

label github 
label food   with color #e74c3c
.EE

.SH EDITING TASKRC

For type \fBtask\fR \fIprimary\fR can be formatted as
\fB@\fIN\fR, where N - number of task in table.

\fBNOTE\fR: if edit file from \fBa\fR bytes to
\fBb\fR bytes, where \fBa > b\fR in end of file will
be writes with \fBa - b - 1\fR zero bytes.

.SS Base Keys

.TP
\fBprimary\fR \- Primary value of query
.br

.TP
\fBpost\fR \- Post value of query
.br
Cannot be used in \fBlabel\fR. In \fBtask\fR it will be
redirected to \fBstatus\fR.

.SS Task Keys

.TP
\fBlabel\fR \- Label of task
.br
Label must been exists. Cannot be change to
none (not implented).

.TP
\fBpriority\fR \- Priority of task
.br
May be if not a number priority will sets to zero.

.TP
\fBstatus\fR \- Completion status
.br
\fB"completed"\fR - completed, another - not completed.

.PP

For edit description use \fBprimary\fR.

.SS Label Keys

.TP
\fBcolor\fR \- Color of label
.br
Must be formatted as \fB"#HHHHHH"\fR.

.PP

For edit name use \fBprimary\fR.

.SH EXAMPLES

.TP
$ \fBtasks use % table diagnostic\fR
.br
Select file \fB~/.taskrc\fR, print table and
show errors.

.TP
$ \fBtasks edit task @2 label github\fR
.br
Change label of task \fB#2\fR to \fBgithub\fR.

.TP
$ \fBtasks edit label cli color '#e4c4c7'\fR
.br
Change color of label \fBcli\fR to \fB#e4c4c7\fR.

.TP
$ \fBtasks delete task @45646\fR
.br
Hmm, if task \fB#45646\fR exists it will be deleted.
Otherwise last task will be deleted.

.SH AUTHOR AND COPYRIGHT

Written by Amchik.

.SH BUGS

Report found bugs here:
<https://github.com/Amchik/tasks/issues>.
Also you can report by email: <am4ik1337@gmail.com>
or <chkrg@ceheki.org>, telegram: \fB@am4ik1337\fR.


tasks.c
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <errno.h>

#include "include/tasks.h"
#include "include/table.h"
#include "include/utils.h"
#include "include/taskrc.h"

#define COMPLETED "✓"

// Creates Label
struct Label createLabel(const char* name, struct color color) {
  struct Label label = {.name = name, .color = color};
  return(label);
}

// Creates Task
struct Task createTask(unsigned int id, const char* description, struct Label label, int priority) {
  struct Task task = {
    .id = id,
    .description = description,
    .label = label,
    .priority = priority,
    .completed = false
  };
  return(task);
}
// Converts Task to struct row
struct row* Tasktorow(struct Task task) {
  struct row* row = malloc(sizeof(struct row));
  row->color = task.label.color;
  row->records = malloc(sizeof(record_t) * 5);
  row->records[4] = 0;
  row->records[0] = task.description;
  row->records[1] = task.label.name;
  row->records[2] = itoa(task.priority);
  row->records[3] = "";
  if (task.completed) row->records[3] = COMPLETED;
  return(row);
}

struct TaskRcContents parseTaskRc(FILE* stream) {
  struct rcfile file = rcparselns(stream, 0);
  // 2 loops: parsing labels then tasks
  struct Label** labels = calloc(file.length, sizeof(void*));
  struct Task**  tasks  = calloc(file.length, sizeof(void*));
  // 1:
  for(int i = 0; i < file.length; i++) {
    struct rcparseresult* res = file.result[i];
    if (!res || res->error.code 
        || strcmp(res->statement.type, "label") != 0) continue;
    struct Label lbl = {
      .name = res->statement.primary,
      .color = {0,0,0}
    };
    if (res->statement.params) {
      const char* _clr = res->statement.params->value;
      if (_clr && _clr[0] == '#') {
        _clr++;
        errno = 0;
        lbl.color = rgbi(
            strtol(_clr, 0, 16)
            );
        if (errno != 0) {
          errno = 0;
          res->error.code = 3;
        }
      }
    }
    labels[i] = malloc(sizeof(lbl));
    memcpy(labels[i], &lbl, sizeof(lbl));
  }
  // 2:
  for (int i = 0; i < file.length; i++) {
    struct rcparseresult* res = file.result[i];
    if (!res || res->error.code 
        || strcmp(res->statement.type, "task") != 0) continue;
    struct Task task = {
      .id = 0,
      .description = res->statement.primary,
      .completed = (res->statement.post 
          && strcmp(res->statement.post, "completed") == 0),
      .label = NOLABEL,
      .priority = 10
    };
    if (res->statement.params) {
      int i = 0;
      struct rcstatementlist* params = res->statement.params;
      while ((params + i)->parameter != 0) {
        if (strcmp((params + i)->parameter, "label") == 0) {
          // find label
          for (int j = 0; j < file.length; j++) {
            if (labels[j] == 0) continue;
            if (strcmp(labels[j]->name, (params + i)->value) == 0) {
              task.label = *labels[j];
              break;
            }
          }
          if (task.label.name == NOLABEL.name) {
            res->error.code = 1;
          }
        } else if (strcmp((params + i)->parameter, "priority") == 0) {
          errno = 0;
          task.priority = strtol((params + i)->value, 0, 10);
          if (errno != 0) {
            errno = 0;
            task.priority = 10;
            res->error.code = 2;
          }
        }
        i++;
      }
    }
    tasks[i] = malloc(sizeof(task));
    memcpy(tasks[i], &task, sizeof(task));
  }
  struct TaskRcContents ans = {
    .tasks = tasks,
    .labels = labels,
    .results = file.result,
    .length = file.length,
  };
  return(ans);
}
